src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -19,46 +18,86 @@ internal static class HPackEncoder
+        public static byte[] EncodeHeader(string name, string value)
         {
-            int i = 0;
-            length = 0;
-
-            if (buffer.Length == 0)
+            byte[] fromPool = null;
+            Span<byte> span = stackalloc byte[256];
+            while (true)
             {
-                return false;
-            }
-
-            buffer[i++] = 0;
+                if (EncodeHeader(name, value, span, out int length))
+                {
+                    byte[] result = span.Slice(0, length).ToArray();
+                    if (fromPool != null)
+                    {
+                        ArrayPool<byte>.Shared.Return(fromPool);
+                    }
+                    return result;
+                }
 
-            if (i == buffer.Length)
-            {
-                return false;
+                int newLength = span.Length * 2;
+                if (fromPool != null)
+                {
+                    ArrayPool<byte>.Shared.Return(fromPool);
+                }
+                span = fromPool = ArrayPool<byte>.Shared.Rent(newLength);
             }
+        }
 
-            if (!EncodeString(name, buffer.Slice(i), out int nameLength, lowercase: true))
+        public static byte[] EncodeHeaderName(string name)
+        {
+            byte[] fromPool = null;
+            Span<byte> span = stackalloc byte[256];
+            while (true)
             {
-                return false;
-            }
+                if (EncodeHeaderName(name, span, out int length))
+                {
+                    byte[] result = span.Slice(0, length).ToArray();
+                    if (fromPool != null)
+                    {
+                        ArrayPool<byte>.Shared.Return(fromPool);
+                    }
+                    return result;
+                }
 
-            i += nameLength;
+                int newLength = span.Length * 2;
+                if (fromPool != null)
+                {
+                    ArrayPool<byte>.Shared.Return(fromPool);
+                }
+                span = fromPool = ArrayPool<byte>.Shared.Rent(newLength);
+            }
+        }
 
-            if (i >= buffer.Length)
+        public static bool EncodeHeader(string name, string value, Span<byte> buffer, out int length)
+        {
+            if (EncodeHeaderName(name, buffer, out int nameLength) &&
+                EncodeHeaderValue(value, buffer.Slice(nameLength), out int valueLength))
             {
-                return false;
+                length = nameLength + valueLength;
+                return true;
             }
 
-            if (!EncodeString(value, buffer.Slice(i), out int valueLength, lowercase: false))
+            length = 0;
+            return false;
+        }
+
+        public static bool EncodeHeaderName(string name, Span<byte> buffer, out int length)
+        {
+            if (buffer.Length > 1 &&
```

Comment:
`if ((uint)buffer.Length > 1)` to eliminate a bounds check in the following `Slice`?

Reply:
Ok